defaulting-capi-clusters/README.md

```diff
@@ -0,0 +1,57 @@
+# Defaulting of CAPI clusters with webhooks
+
+This RFC describes a pattern for defaulting of CAPI clusters by utilizing Kubernetes webhooks extensively.
+
+## Problem statement
+
+We want clusters to be created with defaulted values.
+These values should match the clusters in our [current product reasonably well](https://intranet.giantswarm.io/docs/product/pdr/006_capi-product-guidelines-values/).
+Therefore we will need to default some values differently from upstream Cluster-API.
+
+We do not expect to have full control of the method of cluster creation in the future.
+Therefore we have to account for different methods of cluster creation ( e.g. kubectl, kubectl-gs, happa, gitops, ...) while keeping defaulting consistent between those methods.
+
+## Defaulting with webhooks
+
+Using mutating webhooks allows us to centralize the defaulting logic into webhooks.
+Simplified clients can then rely on defaulted responses from the Kubernetes API by utilizing the server side dry-run feature.
+
+The centralized defaulting logic additionally enables us to have no hardcoded values in any clients.
+
+### Simple defaulting flow
+
+We initially want to have a very simple defaulting flow which allows us to create Kubernetes clusters reliably.
+The cluster creation is focused on kubectl-gs without excluding any other tooling to be used.
+
+This is a high level overview of the cluster creation flow until the Custom Resources are applied:
+1. `User` triggers cluster templating with `kubectl-gs template cluster`
+2. `kubectl-gs` fetches upstream CAPI templates
+3. `kubectl-gs` **removes** values from CAPI templates which we want to default differently
+4. `kubectl-gs` returns minimal templates
+5. `User` either `kubectl apply`s minimal templates or runs `dry-run`
+6. `Admission Webhooks` default all missing values
+
+From this overview we can deduct a set requirements for our admission webhooks:
+- Webhooks must not have any side-effects (e.g. no additional objects are created or deleted)
+- Webhooks must be able to fully default a Custom Resource without context of other Custom Resources in the request (e.g. a `MachineDeployment` must be fully defaulted even if the `Cluster` CR is not known)
```

Comment:
I guess the validation step (after mutation) must make sure the properties are correct. Have you discussed in general how to make that? I see will be similar to defaulting (mutation)

Reply:
Yes - we currently don't focus on it but generally the workflow is the same for now.